### PR TITLE
Engine abstraction alignment spike

### PR DESCRIPTION
**Pull request by pihme**
## Description

Not a PR; only a branch to align on the next steps with all stakeholders. Please feel free to comment here, or submit additional commits to the branch with changes you want to see.

Current focus:
* processing for non-engine records as part of hot backup #9613 
* support for logical transactions #9420 

## Non Engine Records
* Not sure yet, how stream processor will know which records to present to which record processor
* But then again, this is on the platform side, so go crazy :-)
* Did some renaming to support the notion that there might be other record processors than the engine

## Logical transaction
* basically, there is now a `reset()` method in the `ProcessingResultBuilder`. I hope this is enough :thinking:

### CODE REVIEW

File: engine/src/main/java/io/camunda/zeebe/engine/api/ProcessingResultBuilder.java [hunk 1]
```diff
@@ -7,13 +7,48 @@
...
+  ProcessingResultBuilder appendPostCommitTask(Runnable r);
+
   /**
-   * Resets the processing result build to its initial states (removes all follow up records,
-   * responses and commit tasks
+   * Resets the processing result build to its initial states (removes all follow up records, the
+   * response and post commit tasks)
    */
-  void reset();
+  ProcessingResultBuilder reset();
```
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/api/ProcessingResultBuilder.java:51]: This is probably not so important, but it made me wonder so I wrote it down.

❓ Can you clarify what this method does and returns:
- it only returns a new, empty ProcessingResultBuilder without touching the existing ProcessingResultBuilder, so the caller can still access the original data in the original builder object,
- it resets the existing ProcessingResultBuilder but returns the cleared out data, so the cleared out data can be inspected by the caller,
- it resets the existing ProcessingResultBuilder and returns this empty ProcessingResultBuilder, just for chained calls.

**Review by korthout — APPROVED** [review]
Thanks @pihme Nice work!

I think only my second comment adds something worthwhile for the alignment. Please have a look.

File: engine/src/main/java/io/camunda/zeebe/engine/api/ProcessingResultBuilder.java (hunk 1 above)
Comment (pihme) [reply engine/src/main/java/io/camunda/zeebe/engine/api/ProcessingResultBuilder.java:51]: The last option.

File: engine/src/main/java/io/camunda/zeebe/engine/api/ProcessingResultBuilder.java [hunk 2]
```diff
@@ -9,5 +9,11 @@
 
 public interface ProcessingResultBuilder {
 
+  /**
+   * Resets the processing result build to its initial states (removes all follow up records,
+   * responses and commit tasks
+   */
+  void reset();
+
```
[the remark this answers is not in the document]
Comment (pihme) [reply engine/src/main/java/io/camunda/zeebe/engine/api/ProcessingResultBuilder.java:52]: Yes, the StreamProcessor is in charge of the lifecycle of the ProcessingResultBuilder. I think the contract should be that the result builder is empty when it is passed to the engine. So it wouldn't fill it with anything

[the remark this answers is not in the document]
Comment (korthout) [reply engine/src/main/java/io/camunda/zeebe/engine/api/ProcessingResultBuilder.java:52]: ❓ Why not just let the engine create it? That way we don't need such a convention and the API simply restricts it

[the remark this answers is not in the document]
Comment (pihme) [reply engine/src/main/java/io/camunda/zeebe/engine/api/ProcessingResultBuilder.java:52]: That's a good question.

My line of thinking was: 
* The engine implements the `RecordProcessor` interface
* The platform calls the engine and provides implementations for all incoming and outgoing classes of the itnerface
* This way the platform can choose an implementation that is optimal for writing the result back to the stream

In this line of thinking, the engine would use what it is being offered. It would use the processing context to get a hold of the builder (implemented by platform) and then use the builder to generate the processing result (also implemented by the platfom).

At the same time, it falls a bit on its back if the engine is cheeky and just implements the `ProcessingResult` itself. Then it would avoid the nice builder and I think the platform still needs to have a way to handle it. I.e. the platform shouldn't expect that the returned `ProcessingResult` can be cast to a certain class.

So there is a conceptual contract which is not visible in the technical contract.

I will keep your proposal in mind as I work on the refactoring. I see merit in what you are suggesting.

I see a slight benefit in the other approach in terms of testing. If you can mock the processing builder that is passed in and used by the engine, you can observe more than just looking at the returned result. But there we stumble into different philosophies of testing again. 

But even if you just want to observe the output, it would still make it easier to inject the builder into the code and use a different builder in testing. E.g. in production we will probably use a process result which holds a couple of byte buffers, but for testing we could inject a builder which gives me a list of records, and then I can write tests against the test implementation of the process result.

**Comment by deepthidevaki** [comment]
@pihme I have some questions on the PR.

* Not clear who is using `TypedRecordProcessorContext`. Should the `listener(..)` be in the interface? To me it looks like it should be a read-only interface. The "context" will be provided by the StreamProcessor which the "engine" can read.
* Related to above - What is the difference between RecordProcessorContext and TypedRecordProcessorContext? 
* `TypedRecordProcessor` doesn't need to be in the api, right?. It is internal to EngineImpl. StreamProcessor only needs `RecordProcessor`.

File: engine/src/main/java/io/camunda/zeebe/engine/api/ProcessingResultBuilder.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (korthout) [reply engine/src/main/java/io/camunda/zeebe/engine/api/ProcessingResultBuilder.java:52]: This actually makes sense as well 😄  Thanks for that explanation. ❤️ 

💭 If the platform is responsible for an implementation that is optimal for their needs (specific way of writing records and such), then it makes sense that they control the lifecycle. If they have a need to write data into it that cannot be cleared out, then they should deal with that in their implementation of the reset method.

**Comment by pihme** [comment]
@korthout Will remove you from the reviewer list for a time being. I still want to use this branch to align on the next steps, which probably will focus on Deethi for now. Will invite you back in, when there is something for you to see.
